File: src/my_assignments/turtlebot3/zk_turtlebot3_move_PID.cpp
#include "ros/ros.h"
#include "sensor_msgs/LaserScan.h"
#include "laserscan/LaserScanner.h"
#include "geometry_msgs/Twist.h"
#include "std_srvs/Empty.h"
#include "utility_lib.h"
#include <cstdlib>

using namespace std;

int angToIdx(double angle, double angle_inc){
  if (angle<0){
      angle = 360.0 + angle;
  }
  return (int)(angle/radian2degree(angle_inc));
};

void getMinRange(double &res, sensor_msgs::LaserScan & laser_scan_msg, double angle, double delta_ang){
  int index0 {angToIdx(angle-delta_ang, laser_scan_msg.angle_increment)};
  int index1 {angToIdx(angle+delta_ang, laser_scan_msg.angle_increment)};
  if (index1>index0){
    res = LaserScanner::getAverageRange(laser_scan_msg, index0, index1);
  }else{
    double rs1 {0.0}, rs2 {0.0};
    rs1 = LaserScanner::getAverageRange(laser_scan_msg, 0, index1);
    rs2 = LaserScanner::getAverageRange(laser_scan_msg, index0, laser_scan_msg.ranges.size());
    res = (rs1<rs2) ? rs1 : rs2;
  } 

  if(std::isnan(res))
    res = 100;
};

class Listener{
public:
  double dist_ahead {100.0};
  double dist_left {100.0};
  double dist_right {10.0};
public:
  void scanCallback(sensor_msgs::LaserScan scanMessage){
    getMinRange(this->dist_ahead, scanMessage, 0, 45);
    getMinRange(this->dist_left, scanMessage, 45, 20);
    getMinRange(this->dist_right, scanMessage, -45, 20);
  };
};

void resetSpeedCmd(geometry_msgs::Twist &vel_msg){
  vel_msg.linear.x =0.0;
  vel_msg.angular.z =0.0;
};

ros::Subscriber scanSubscriber;
ros::Publisher velocity_publisher;

//void scanCallback (sensor_msgs::LaserScan scanMessage, distances &dist);
int main(int argc, char **argv)
{
  ros::init(argc, argv, "turtlebot3_move_PID");
  ros::NodeHandle n;

  // RESET THE SIMULATION
  ros::ServiceClient client = n.serviceClient<std_srvs::Empty>("/gazebo/reset_simulation");
  std_srvs::Empty req;
  cout << "Reseting the simulator" << endl;
  client.call(req);

  // Subscribe to the scan topic
  Listener listener;
  scanSubscriber = n.subscribe("/scan", 10, &Listener::scanCallback, &listener);
  velocity_publisher = n.advertise<geometry_msgs::Twist>("/cmd_vel", 1000);

  geometry_msgs::Twist vel_msg;
  ros::Rate loop_rate(100);

  int stop_cnts {0};
  int state_id {0};
  int rot_sign {1};
  const double ahead_th {4}, lat_dist {0.4};
  double t0 = ros::Time::now().toSec();
  double t1 = ros::Time::now().toSec();  
  do{

    vel_msg.linear.x = (listener.dist_ahead > ahead_th) ? 0.4 : 0.4 - (ahead_th-listener.dist_ahead)*0.1;
    double delta_lat_dist {listener.dist_left - listener.dist_right};
    vel_msg.angular.z = (delta_lat_dist <-lat_dist or delta_lat_dist > lat_dist) ? (delta_lat_dist - lat_dist)*0.001 : 0.0;

    //vel_msg.linear.x = (vel_msg.linear.x<.8) ? vel_msg.linear.x : 0.8;  
    vel_msg.angular.z = (vel_msg.angular.z<-.5) ? -.5 : ((vel_msg.angular.z>.5) ? 0.5 : vel_msg.angular.z);  
    velocity_publisher.publish(vel_msg);

    cout << vel_msg.linear.x << " " << vel_msg.angular.z << endl;

    ros::spinOnce();
    loop_rate.sleep();
    t1 = ros::Time::now().toSec();    
  }while((t1-t0)<10);

  resetSpeedCmd(vel_msg);
  velocity_publisher.publish(vel_msg);
  return 0;
}

File: src/my_assignments/turtlebot3/zk_turtlebot3_move_PID.py
#!/usr/bin/env python
# rosrun ros_essentials_cpp turtlesim_cleaner.py
import rospy
import math
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose
from std_srvs.srv import Empty
from sensor_msgs.msg import LaserScan


def min_dist_in_range(ranges, rg, angle_increment):
    range_init = int(math.radians(rg[0]%360)//angle_increment)
    range_end = int(math.radians(rg[1]%360)//angle_increment)
    if range_end > range_init:
        sector_data = ranges[range_init:range_end]
    else:
        sector_data = ranges[range_init:] + ranges[0:range_end]
    #print(sector_data)
    return min(sector_data)

def scan_callback(msg, data):
    
    data['dist_ahead'] = min_dist_in_range(msg.ranges, [-6, 6], msg.angle_increment)
    data['dist_left'] = min_dist_in_range(msg.ranges, [0, 30], msg.angle_increment)
    data['dist_right'] = min_dist_in_range(msg.ranges, [-30, 0], msg.angle_increment)
    #print(data)

def move_foward(speed_cmd, pub, data_dct, rate):
    done = False
    speed_cmd.linear.x = 0.3

    while not done:
        done = data_dct['dist_ahead']<0.6
        diff = data_dct['dist_left'] - data_dct['dist_right']
        if abs(diff)>.8:
            speed_cmd.angular.z = (diff)*0.3
        else:
            speed_cmd.angular.z = max([0, (1.5-data_dct['dist_ahead'])])*0.8

        speed_cmd.linear.x = 0.3 - max([0, (1.5-data_dct['dist_ahead'])]) * 0.3
        print(data_dct, speed_cmd.angular.z)
        pub.publish(speed_cmd)
        rate.sleep()
        #print(data_dct['dist_ahead'], speed_cmd.linear.x)
    speed_cmd.linear.x = 0
    speed_cmd.angular.z = 0
    pub.publish(speed_cmd)

def rotate_until_free(speed_cmd, pub, data_dct, rate):
    done = False
    speed_cmd.angular.z = 0.5

    while not done:
        done = data_dct['dist_ahead']>3
        pub.publish(speed_cmd)
        rate.sleep()
        #print(data_dct['dist_ahead'], speed_cmd.angular.z)
    speed_cmd.angular.z = 0.0
    pub.publish(speed_cmd)


if __name__ == '__main__':
    rospy.init_node('turtlebot3_move_while_possible')

    # Reset the turtle simulator
    print("Reseting the simulator")
    reset_turtle = rospy.ServiceProxy('/gazebo/reset_simulation', Empty)
    reset_turtle()

    # create the command publisher and the pose subscriber
    data_dct = {'dist_ahead':100, 'dist_left':100, 'dist_right':100}
    rospy.Subscriber("/scan", LaserScan, scan_callback, data_dct)
    pub = rospy.Publisher('/cmd_vel', Twist, queue_size=10)

    # Initialize the speed message
    speed_cmd = Twist()
    

    # # Set the frequency at which we wwant to iterate
    rate = rospy.Rate(20) # 1hz

    while True:
        #x = max([0, ])
        move_foward(speed_cmd, pub, data_dct, rate)
        rotate_until_free(speed_cmd, pub, data_dct, rate)
    